cloudformation/curprocessor-sam.yml

```diff
@@ -2,17 +2,147 @@ AWSTemplateFormatVersion : '2010-09-09'
+
+  CostUsageReport:
+    Type: AWS::S3::Bucket
+    Properties: 
+      AccessControl: Private
+      BucketName: !Ref BucketName
+      
+  CostUsageReportBucketPolicy:
+    Type: AWS::S3::BucketPolicy
+    Properties: 
+      Bucket: !Ref CostUsageReport
+      PolicyDocument: 
+        Id: AllowLogDelivery
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Principal: 
+              AWS: "arn:aws:iam::386209384616:root" 
+            Resource: !Sub ${CostUsageReport.Arn}
+            Action:
+              - s3:GetBucketAcl
+              - s3:GetBucketPolicy
+          - Effect: Allow
+            Principal: 
+              AWS: "arn:aws:iam::386209384616:root" 
+            Resource: 
+              - !Sub ${CostUsageReport.Arn}/${ReportPathPrefix}*
+              - !Sub ${CostUsageReport.Arn}/aws-programmatic-access-test-object
+            Action: s3:PutObject
+     
+  StepFunctionsRole:
+    Type: AWS::IAM::Role
+    Properties:
+      Path: /
+      AssumeRolePolicyDocument:
+        Version: 2012-10-17
+        Statement:
+          - Effect: Allow
+            Action: sts:AssumeRole
+            Principal:
+              Service: !Sub states.${AWS::Region}.amazonaws.com
+      Policies:
+        - PolicyName: InvokeLambdaFunctions
+          PolicyDocument:
+            Statement:
+              - Effect: Allow
+                Resource:
+                  - !GetAtt ProcessCostUsageFunction.Arn
+                  - !GetAtt CreateAthenaResourcesFunction.Arn
+                  - !GetAtt UpdateMetadataFunction.Arn
+                  - !GetAtt InitializeAthenaQueriesFunction.Arn
+                Action:
+                  - lambda:InvokeFunction
+
+  StepFunctions:
+    Type: AWS::StepFunctions::StateMachine
+    Properties:
+      StateMachineName: AWSCostUsageReport 
+      RoleArn: !GetAtt StepFunctionsRole.Arn
+      DefinitionString: !Sub |
+          {
+            "Comment": "Process AWS Cost and Usage reports and prepare them for Athena.",
+            "StartAt": "ProcessCostUsageReport",
+            "States": {
+              "ProcessCostUsageReport": {
+                "Type": "Task",
+                "Resource": "${ProcessCostUsageFunction.Arn}",
+                "Next": "CreateAthenaResources"
+              },
+              "CreateAthenaResources": {
+                "Type": "Task",
+                "Resource": "${CreateAthenaResourcesFunction.Arn}",
+                "Next": "UpdateMetadata"
+              },
+              "UpdateMetadata": {
+                "Type" : "Task",
+                "Resource": "${UpdateMetadataFunction.Arn}",
+                "Next": "InitializeAthenaQueries"
+              },
+              "InitializeAthenaQueries": {
+                "Type" : "Task",
+                "Resource": "${InitializeAthenaQueriesFunction.Arn}",
+                "End": true
+              }
+            }
+          }
+
+  xAcctStepFunctionStarterLogGroup:
```

Comment:
One issue with having log groups created in the template is that it breaks compatibility with existing stacks. It's not possible to update an existing CloudFormation stack with the new template, since a log group already exists with the same name and CloudFormation aborts the stack update. One workaround would be to delete existing log groups, but that would mean losing log data for existing functions, which is not really an option. Perhaps making Log Group conditional, but it feels like log group creation in the template is not necessary.

Reply:
So the reason why i always create LogGroups from the CFN template is that you have control on the cleanup path. If you delete your stack your logging is removed which is useful for dev and test for prod you are not deleting the stack unless you discontinue the service in that case and you want to retain the logs you can archive it to glacier or set a delete policy to retain the log group.

An option is to rename the resource id's of the functions that way an update should work but the template as it is now you first need to deploy it then fill in the values in a local file and then redeploy. Took some effort for me to set this up (hence the reason why i created the PR to make it easier for other people to give this a try) so do you have any idea how many deployments we are talking about?

Or i can delete the log group resources, your project your call.

Reply:
How about adding a condition, something like `CreateLambdaFunctionLogGroups` that defaults to 'Enabled'? This way new stack creations will work, but updates to existing ones can set it to 'Disabled'.

Reply:
That's actually a better solution, i have pushed it to the PR.